### PR TITLE
Added 1-20% jitter to expiration value stored by AwsCredentials struc…

### PR DESCRIPTION
**Pull request by jdelapla**
…t. This is part of a larger effort to help customers with large fleets distribute the timings of their attempts to reconnect to our servers.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by MushMal** [comment]
PIC already has a jitter solution. If the jitter is intended for WebRTC then use the signaling client itself for jitter and not the actual provider.

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/AuthIntegration.c#L116

Follow the same solution

**Comment by jdelapla** [comment]
A maximum of a 3 minute jitter does not scale well for users with a large fleet.

File: src/source/Common/IotCredentialProvider.c [hunk 1]
```diff
@@ -176,10 +176,20 @@ STATUS parseIotResponse(PIotCredentialProvider pIotCredentialProvider, PCallInfo
...
 
-    // Fix-up the expiration to be no more than max enforced token rotation to avoid extra token rotations
-    // as we are caching the returned value which is likely to be an hour but we are enforcing max
-    // rotation to be more frequent.
-    expiration = MIN(expiration, currentTime + MAX_ENFORCED_TOKEN_EXPIRATION_DURATION);
+    // Need to have large numbers since all items are in terms of HUNDREDS_NANOS_IN_A_SECOND. -1 to round down
+    randMultiplier = (ULONG_MAX / RAND_MAX) - 1;
+
+    // add randomized jitter between 1-20% of expiration
+    srand(currentTime / ((UINT64) sessionToken[rand() % (sessionTokenLen - 1)] + (UINT64) sessionToken[rand() % (sessionTokenLen - 1)]));
```
Comment (hassanctech) [inline src/source/Common/IotCredentialProvider.c:183]: Can (UINT64) sessionToken[rand() % (sessionTokenLen - 1)] be 0?

Comment (jdelapla) [reply src/source/Common/IotCredentialProvider.c:183]: No. It's a JSON string, not a binary data, being converting to a cstring, the -1 on the length prevents the null terminating character at the end from being included.

Comment (jdelapla) [reply src/source/Common/IotCredentialProvider.c:183]: No it's a JSON string, and JSON string's can't contain null terminating characters in the middle of the string. The last character is omitted with the -1.
https://docs.aws.amazon.com/STS/latest/APIReference/API_Credentials.html

API showing it's a JSON string returned.

File: src/source/Common/IotCredentialProvider.c [hunk 2]
```diff
@@ -176,10 +176,20 @@ STATUS parseIotResponse(PIotCredentialProvider pIotCredentialProvider, PCallInfo
         pIotCredentialProvider->pAwsCredentials = NULL;
     }
 
-    // Fix-up the expiration to be no more than max enforced token rotation to avoid extra token rotations
-    // as we are caching the returned value which is likely to be an hour but we are enforcing max
-    // rotation to be more frequent.
-    expiration = MIN(expiration, currentTime + MAX_ENFORCED_TOKEN_EXPIRATION_DURATION);
```
Comment (hassanctech) [inline src/source/Common/IotCredentialProvider.c:182]: Don't we still need this?

**Review by hassanctech — APPROVED** [review]
lgtm

**Pull request merged**